roachprod: test selector

nameisbhaskar opened the pull request:
The test selector code responsible for running snowflake query and select the tests based on the criteria mentioned in https://cockroachlabs.atlassian.net/wiki/spaces/~7120207825326fb5e546c194029506f2c5335e/pages/3596091522/Test+Selection+Criteria

Informs: #119630
Epic: none

pav-kv committed:
raft: consolidate all append message sending

This PR consolidates all decision-making about sending append messages
into a single maybeSendAppend method. Previously, the behaviour depended
on the sendIfEmpty flag which was set/unset depending on the context in
which the method is called. This is unnecessary because the Progress
struct contains enough information about the leader->follower flow
state, so maybeSendAppend can be made stand-alone.

In follow-up PRs, the consolidated maybeSendAppend method will be used
to implement a more flexible message flow control.

Epic: CRDB-37515
Release note: none

stevendanna committed:
jobs,rttanalysis: benchmark jobs with stats

Job system queries are rather sensitive to the presence of stats. For
example, in the benchmarks already present in this file, we see a
substantial improvement when stats are present:

```
Jobs/crdb_internal.system_jobs

+--------------+--------------+------------------------+
| HEAD^1       | HEAD         |                        |
|--------------+--------------+------------------------|
| sec/op       | sec/op       | vs base                |
| 160.90m ± 4% | 73.07m ± 9%  | -54.59% (p=0.000 n=10) |
|--------------+--------------+------------------------|
| roundtrips   | roundtrips   | vs base                |
| 8.000 ± 0%   | 3.000 ± 0%   | -62.50% (p=0.000 n=10) |
|--------------+--------------+------------------------|
| B/op         | B/op         | vs base                |
| 57.99Mi ± 2% | 26.41Mi ± 1% | -54.46% (p=0.000 n=10) |
|--------------+--------------+------------------------|
| allocs/op    | allocs/op    | vs base                |
| 899.2k ± 2%  | 315.2k ± 1%  | -64.95% (p=0.000 n=10) |
+--------------+--------------+------------------------+
```

This has implications for job system queries on 23.1 and 23.2 where we
do not necessarily have automatic stats for system.jobs because of

Epic: none
Release note: None

herkolategan committed:
roachprod: pass same set of keys with session

Previously, the public key list used by session diverged from the list specified
in the config `defaultPubKeyNames`. This change makes the list public and reuses
it for remote sessions as well.

Release Note: None
Epic: None

herkolategan committed:
scripts: drtprod fix dns

Previously, clusters were not secure by default. The script has been updated to
handle "HTTPS" when parsing a result to create DNS.

Release Note: None
Epic: None

rail committed:
release: automatically detect if version is latest

Previously, we relied on manually set variable to distinguish latest
versions. This has been error-prone.

This PR uses the releases DB to detect if the current version is latest.

Epic: none
Release note: None

michae2 committed:
sql: add defensive type check to sql.(*optTableStat).init

In statistics builder we assume that the datums decoded from histogram
upper bounds are comparable with datums in spans and constraints. If the
histogram column type were ever different from the table column type,
however, this assumption would not hold.

This should never happen, because histograms are associated with a
column ID, and ALTER TABLE should never re-use a column ID during ALTER
TABLE ADD COLUMN or ALTER TABLE ALTER COLUMN TYPE. But just to be
defensive, add a check to sql.(*optTableStat).init that skips over the
TableStatistic if the histogram column type isn't equivalent to the
table column type.

Also add this same typecheck to ALTER TABLE INJECT STATISTICS to guard
against injecting histograms that don't match the column type. As part
of this fix I'm removing the testcase added in #46552 which deliberately
injects statistics with a histogram of a different type.

Informs: #124181

Release note: None

michae2 committed:
sql/stats: evict stats cache entry if user-defined types have changed

When adding table statistics to the stats cache, we decode histogram
upper bounds into datums. If the histogram column uses a user-defined
type, we hydrate the type and use this to decode.

In statistics builder, these histogram upper bound datums are compared
against datums in spans and constraints. The comparisons assume that the
datums are of equivalent type, but if the user-defined type has changed
sometime after loading the stats cache entry, this might not be true.

If the user-defined type has changed, we need to evict and re-load the
stats cache entry so that we decode histogram datums with a freshly-
hydrated type.

(We were already checking UDT versions when building the optTable in
sql.(*optCatalog).dataSourceForTable, but the newly-built optTable used
the existing table statistics instead of refreshing those, too.)

Fixes: #124181

Release note (bug fix): Fix a bug where a change to a user-defined type
could cause queries against tables using that type to fail with an error
message like:

  "histogram.go:694: span must be fully contained in the bucket"

The change to the user-defined type could come directly from an ALTER
TYPE statement, or indirectly from an ALTER DATABASE ADD REGION or DROP
REGION statement (which implicitly change the crdb_internal_region
type).

This bug has existed since UDTs were introduced in v20.2.

michae2 committed:
sem/tree: check oid and version in tree.(*DEnum).CompareError

Make sure when we're comparing two enum datums that they are, in fact,
the same enum type.

Informs: #124181

Release note: None

nameisbhaskar committed:
roachprod: delete cluster config on GC

currently, prometheus cluster configs are not deleted on
gc. This makes stale entries to remain. This PR deletes the
cluster configs on GC.

There is another minor fix for promhelpers access. The access
to cloud storage was failing due to missing credentials.

Fixes: #124599
Epic: none

nameisbhaskar committed:
roachprod: test selector

The test selector code responsible for running snowflake
query and select the tests based on the criteria mentioned
in https://cockroachlabs.atlassian.net/wiki/spaces/~7120207825326fb5e546c194029506f2c5335e/pages/3596091522/Test+Selection+Criteria

Informs: #119630
Epic: none

Commit:
roachprod: fix check for errors in `find-ports`

When finding ports, we only check for the `err` return value from
`runCmdOnSingleNode`. However, that function only returns an error if
there is a string expansion error. For command errors, the caller
needs to check `res.Err`, which was not happening before.

Informs: #124715

pav-kv committed:
tracker: rename MsgAppFlowPaused

Epic: none
Release note: none

rafiss committed:
schemachanger_test: increase shard count

Release note: None

cockroach-teamcity committed:
master: Update pkg/testutils/release/cockroach_releases.yaml

Update pkg/testutils/release/cockroach_releases.yaml with recent values.

Epic: None
Release note: None
Release justification: test-only updates

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/124793)
<!-- Reviewable:end -->

stevendanna committed:
streamingccl: remove unused function

This function is no longer used now that we pass the full checkpoint
data rather than just the frontier timestamp.

Epic: none
Release note: None
